pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs

```diff
@@ -0,0 +1,190 @@
+//! IsmpDispatcher precompiles for pallet-evm
+
+use crate::{
+    dispatcher::Dispatcher,
+    evm::abi::{
+        DispatchGet as SolDispatchGet, DispatchPost as SolDispatchPost,
+        PostResponse as SolPostResponse,
+    },
+    weight_info::WeightInfo,
+    Config, GasLimits, Pallet,
+};
+use alloc::str::FromStr;
+use alloy_sol_types::SolType;
+use core::marker::PhantomData;
+use fp_evm::{
+    ExitError, ExitSucceed, Precompile, PrecompileFailure, PrecompileHandle, PrecompileOutput,
+    PrecompileResult,
+};
+use frame_support::traits::Get;
+use ismp_rs::{
+    host::StateMachine,
+    router::{DispatchGet, DispatchPost, DispatchRequest, IsmpDispatcher, Post, PostResponse},
+};
+use pallet_evm::GasWeightMapping;
+use sp_core::{H256, U256};
+
+/// Ismp Request Dispatcher precompile for evm contracts
+pub struct IsmpPostDispatcher<T> {
+    _marker: PhantomData<T>,
+}
+
+impl<T> Precompile for IsmpPostDispatcher<T>
+where
+    T: Config + pallet_evm::Config,
+    <T as frame_system::Config>::Hash: From<H256>,
+{
+    fn execute(handle: &mut impl PrecompileHandle) -> PrecompileResult {
+        let input = handle.input();
+        let context = handle.context();
+        let weight = <T as Config>::WeightInfo::dispatch_post_request();
+
+        // The cost of a dispatch is the weight of calling the dispatcher plus an extra storage read
+        // and write
+        let cost = <T as pallet_evm::Config>::GasWeightMapping::weight_to_gas(
+            weight.saturating_add(<T as frame_system::Config>::DbWeight::get().reads_writes(1, 1)),
+        );
+
+        let dispatcher = Dispatcher::<T>::default();
+        let post_dispatch =
+            SolDispatchPost::decode(input, true).map_err(|e| PrecompileFailure::Error {
+                exit_status: ExitError::Other(format!("Failed to decode input: {:?}", e).into()),
+            })?;
+        let gas_limit = post_dispatch.gasLimit;
+        let post_dispatch = DispatchPost {
+            dest: parse_state_machine(post_dispatch.dest)?,
+            from: context.caller.0.to_vec(),
+            to: post_dispatch.to,
+            timeout_timestamp: u256_to_u64(post_dispatch.timeoutTimestamp)?,
+            data: post_dispatch.data,
+        };
+        handle.record_cost(cost)?;
+        match dispatcher.dispatch_request(DispatchRequest::Post(post_dispatch)) {
+            Ok(_) => {
+                let nonce = Pallet::<T>::previous_nonce();
+                GasLimits::<T>::insert(nonce, gas_limit);
```

Comment:
but this gas limit is for the destination chain

Reply:
The gas limit for the destination chain is encoded with the request data as part of the `ContractData` struct.